providers/common/macs/cmac.c

```diff
@@ -0,0 +1,210 @@
+
+    if (macctx != NULL) {
+        CMAC_CTX_free(macctx->ctx);
+        EVP_CIPHER_meth_free(macctx->alloc_cipher);
+        OPENSSL_free(macctx);
+    }
+}
+
+static int cmac_copy(void *vdst, void *vsrc)
+{
+    struct cmac_data_st *dst = vdst;
+    struct cmac_data_st *src = vsrc;
+
+    if (!CMAC_CTX_copy(dst->ctx, src->ctx))
+        return 0;
+
+    if (src->alloc_cipher != NULL && !EVP_CIPHER_upref(src->alloc_cipher)) {
+        CMAC_CTX_free(dst->ctx);
+        return 0;
+    }
+
+    dst->tmpengine = src->tmpengine;
+    dst->tmpcipher = src->tmpcipher;
+    dst->alloc_cipher = src->alloc_cipher;
+    return 1;
+}
+
+static size_t cmac_size(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return EVP_CIPHER_CTX_block_size(CMAC_CTX_get0_cipher_ctx(macctx->ctx));
+}
+
+static int cmac_init(void *vmacctx)
+{
+    struct cmac_data_st *macctx = vmacctx;
+    int rv = CMAC_Init(macctx->ctx, NULL, 0, macctx->tmpcipher,
+                       (ENGINE *)macctx->tmpengine);
+
+    macctx->tmpcipher = NULL;
+    macctx->tmpengine = NULL;
+
+    return rv;
+}
+
+static int cmac_update(void *vmacctx, const unsigned char *data,
+                       size_t datalen)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return CMAC_Update(macctx->ctx, data, datalen);
+}
+
+static int cmac_final(void *vmacctx, unsigned char *out, size_t *outl,
+                      size_t outsize)
+{
+    struct cmac_data_st *macctx = vmacctx;
+
+    return CMAC_Final(macctx->ctx, out, outl);
+}
+
+static const OSSL_PARAM known_params[] = {
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_ALGORITHM, NULL, 0),
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_CIPHER, NULL, 0),     /* Same as "algorithm" */
+    OSSL_PARAM_utf8_string(OSSL_MAC_PARAM_PROPERTIES, NULL, 0), /* instead of ENGINE */
+    OSSL_PARAM_octet_string(OSSL_MAC_PARAM_KEY, NULL, 0),
+    OSSL_PARAM_END
+};
+static const OSSL_PARAM *cmac_set_param_types(void)
+{
+    return known_params;
+}
+
+/*
+ * ALL parameters should be set before init().
+ */
+static int cmac_set_params(void *vmacctx, const OSSL_PARAM params[])
+{
+    struct cmac_data_st *macctx = vmacctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_MAC_PARAM_CIPHER)) != NULL
+        || (p = OSSL_PARAM_locate(params, OSSL_MAC_PARAM_ALGORITHM)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+
+        {
+            const char *algoname = p->data;
+            const char *properties = NULL;
+
+            if ((p = OSSL_PARAM_locate(params, OSSL_MAC_PARAM_PROPERTIES)) != NULL) {
+                if (p->data_type != OSSL_PARAM_UTF8_STRING)
+                    return 0;
+
+                properties = p->data;
+            }
+
+            if (macctx->alloc_cipher != NULL)
+                EVP_CIPHER_meth_free(macctx->alloc_cipher);
+
+            if ((macctx->tmpcipher = macctx->alloc_cipher =
+                 EVP_CIPHER_fetch(NULL, algoname, properties)) == NULL) {
```

Comment:
`libctx` should not be NULL. This is a bit of a problem actually. In the FIPS provider it should be the provider context that was passed as the parameter to `cmac_new` (which is really an OPENSSL_CTX). But in the default provider we should use the OPENSSL_CTX from the libcrypto side. Which makes me realise that this FIPS_MODE only code should perhaps apply to any built-in provider (or perhaps just all providers):

https://github.com/openssl/openssl/blob/85e0ad9f1309afceaa8d3946c9273e089d3c25b8/crypto/provider_core.c#L402-L424

In other words in the default provider, the provider context should be the same object as the "real" OPENSSL_CTX.

Reply:
I would say that any provider is free to use whatever context and whatever properties they want, so let's not push this on *everyone*.

The idea with the provider context is that the provider decides what it contains, not libcrypto, so forcing that to be a `OPENSSL_CTX` at all times is not a good idea either.

Now, what *we* choose to do with *our* providers is a different matter, and of course we can make whatever we choose to use available for others as well, but that's the point, isn't it, that it should remain a choice?
Also, I'm apprehensive of "too magic", so I'm wondering if things can get redesigned so the provider can *ask* for the library context that it's running under...  as an upcall with the `OSSL_PROVIDER` as argument, maybe?

Reply:
Also, note that this PR is a draft, so some things are obviously lacking, and I wanted to see that the code would build at all, and even test.

Reply:
> I would say that any provider is free to use whatever context and whatever properties they want, so let's not push this on everyone.

> The idea with the provider context is that the provider decides what it contains, not libcrypto, so forcing that to be a OPENSSL_CTX at all times is not a good idea either.

I'm definitely not suggesting that. My only suggestion was what the provctx value is by default. Any provider would be free to set it to whatever they want. This is a `void *` so there is nothing forcing an OPENSSL_CTX to be used. Currently it defaults to NULL, except in the special case of the built-in internal recursive FIPS provider where it defaults to point to the "current" OPENSSL_CTX. I'm just saying we can make that the default everywhere.

> Also, I'm apprehensive of "too magic", so I'm wondering if things can get redesigned so the provider can ask for the library context that it's running under... as an upcall with the OSSL_PROVIDER as argument, maybe?

Yes, that could work too.

Reply:
> Yes, that could work too.

Cool, I'll look into that next week.

Reply:
I have changed them all to use the library context that the provider works in.